### DISCUSSION TITLE
Exact frequencies of notes

### DISCUSSION THREAD

**Discussion by andersmattson** [pschatzmann/arduino-audio-tools#664] [Q&A] [answered]
Hi! Just found this beautiful lib when working on building my own (sort of) synth, but I was just wondering about the notes and their frequencies when using the sine wave generator. I've had problems myself generating waveforms in the correct pitch when dabbling with this on my own, generating some "interesting" wobbles when playing two notes at the same time for instance. 

I read an interesting article (can't find it again unfortunately) about a case study where the participants could perceive a difference in frequency down to about 0.3% in the mid-range audible frequencies. So, how exact are the frequencies generated for different notes? Are they down to below where you can hear if they're out of tune?

**Comment by pschatzmann** [comment, marked as answer]
That's an interesting question and I never looked into it. I just recently changed the frequency of the notes from integers to float, so it should be better then in the beginning of the project. I would also expect that it might depend on your [selected generator implementation](https://pschatzmann.github.io/arduino-audio-tools/group__generator.html).

The SineWaveGenerator should be the most exact, but not the fastest implementation because it relies on the sine() function.

It would be cool if you could spend some time in analyzing this...

**Comment by andersmattson** [comment]
I'm already looking in to it, but got stuck on the first try :P

I'm trying to use the Synthesizer to generate two sine waves at the same time, but for some reason when running the code below it still outputs a clean sine wave (no mix of two different ones) and also the frequency is about twice that of the last set note.

I started poking around and my only guess is that since the `DefaultSynthesizerChannel` class is using a static `SineWaveGenerator` it gets updated twice for every sample? No idea how to get around that though.

```
#include "AudioTools.h"

uint16_t sample_rate=44100;
uint8_t channels = 2;

Synthesizer synthesizer;
GeneratedSoundStream<int16_t> in(synthesizer); 
AnalogAudioStream out; 
StreamCopy copier(out, in);

void setup(void) {  

  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Debug);

  // Synthesizer
  auto cfg = in.defaultConfig();
  cfg.channels = channels;
  cfg.sample_rate = 44100;
  in.begin(cfg);

  // Analog output
  auto config = out.defaultConfig(TX_MODE);
  config.sample_rate = sample_rate; 
  out.begin(config);

  synthesizer.keyOn(N_A3, 1.0f);
  synthesizer.keyOn(N_E4, 1.0f);

  Serial.println("started...");
}

void loop() {
  copier.copy();
}
```

**Comment by pschatzmann** [comment]
Oh, I haven't looked at that class for a long time and there were many changes after I created it. I will have a look on Monday.
In the meantime you can start with the sketch described in the Mixing Wiki
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-generator_inputmixer-audiokit/streams-generator_inputmixer-audiokit.ino

**Comment by HobmanX** [comment]
Sounds like this may be similar problem to what I had a few months ago.... see link below....
https://github.com/pschatzmann/arduino-audio-tools/discussions/556

At that time I was trying with an ESP32 PICO KIT using the sound out on internal DAC pins 25 and 26....... never got a solution always get beats when playing 2 sounds that should sound in tune eg playing a sound and its octave or playing a sound and another one one say 4 times higher.
I have now  tried the ESP32 PICO KIT with external I2S DAC  PT8211 and now get a better quality of sound but the beats at so frequency mixes are still there..??.  So would still be interested if a solution can be found.

**Reply by pschatzmann** [reply]
I just tried your sketch: The issue comes from the logger: just change to AudioLogger::Wanring and it will be much better!
